### PR TITLE
feat(core): ai brief - improved logic

### PR DESCRIPTION
**Pull request by RamilGaripov**
refs. metriport/metriport-internal#2510

### Description
- Filtering the input bundles for AI brief generation to keep only the important data prior to sending it to LLM
- Adding langchain for a complete summary generation
- Baking the new logic into the flow for the consolidated query

### Testing

- Local
  - [ ] Tested locally with a utils scripts
    - [ ] Make sure the AI brief is generated and looks reasonable
    - [ ] Check that the brief looks good on the MR summary
- Staging
  - [ ] Generate a brief on a staging org
    - [ ] Make sure to add FF 
    - [ ] Try to leave feedback 

### Release Plan

- [ ] Merge this

### CODE REVIEW

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 1]
```diff
@@ -1,128 +1,747 @@
...
+
+  // this is the prompt for combining the summaries into a single paragraph
+  const summaryTemplateRefined = `
+${systemPrompt}
+
+Today's date is ${todaysDate}.
+Your goal is to write a summary of the patient's most recent medical history, so that another doctor can understand the patient's medical history to be able to treat them effectively.
+Here are the previous summaries written by you of sections of the patient's medical history:
+--------
+{text}
```
Comment (Goncharo) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:126]: ditto use the const mentioned above

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 2]
```diff
@@ -1,128 +1,747 @@
...
+
+SUMMARY:
+`;
+  const SUMMARY_PROMPT_REFINED = PromptTemplate.fromTemplate(summaryTemplateRefined);
+  const summaryChainRefined = new StuffDocumentsChain({
+    llmChain: new LLMChain({
+      llm: llmSummary,
+      prompt: SUMMARY_PROMPT_REFINED,
+    }),
+    documentVariableName: "text",
```
Comment (Goncharo) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:141]: can just be `documentVariableName` if using the const

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 3]
```diff
@@ -1,128 +1,747 @@
...
+  });
+
+  const mapReduce = new MapReduceDocumentsChain({
+    llmChain: summaryChain,
+    combineDocumentChain: summaryChainRefined,
+    documentVariableName: "text",
```
Comment (Goncharo) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:147]: can just be `documentVariableName` if using the const

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 4]
```diff
@@ -1,128 +1,747 @@
...
+  bundle: Bundle<Resource>
+): Promise<string | undefined> {
+  // filter out historical data
+  const numHistoricalYears = 2;
+  const dateFrom = dayjs().subtract(numHistoricalYears, "year").format(ISO_DATE);
+  const filteredBundle = filterBundleByDate(bundle, dateFrom);
+
+  const inputString = prepareBundleForBrief(filteredBundle);
+
+  // TODO: experiment with different splitters
```
Comment (Goncharo) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:74]: tie TODOs with an issue #

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 5]
```diff
@@ -1,128 +1,747 @@
...
+    if (name) {
+      updRes.name = name;
+      if (name.includes("no data")) return undefined;
+    }
+
+    delete updRes.code;
+
+    if (updRes.status === "") delete updRes.status;
+
+    delete res.reasonCode; // TODO: Introduce term server lookup here
```
Comment (Goncharo) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:371]: tie TODOs with an issue #

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts [hunk 6]
```diff
@@ -1,128 +1,747 @@
...
+    if (res.referenceRange) {
+      updRes.referenceRange = res.referenceRange.map(range => {
+        const low = getQuantityString(range.low);
+        const high = getQuantityString(range.high);
+        return {
+          low,
+          high,
+        };
+      });
+    }
```
Comment (Goncharo) [inline packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:426]: nit: a summary of what this is doing would be really nice, and/or break this into specific functions per-resrouce

a single 150 LOC function with no comments will make a dev want to blow their brains out from a maintainability perspective

File: packages/utils/src/customer-requests/medical-record-brief-input.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+// Update this to staging or local URL if you want to test the brief link
+const dashUrl = "http://dash.metriport.com";
+
+const SOURCE_DIR = "";
+const SOURCE_PATIENT_ID = "";
+const SOURCE_BUNDLE_FILE = `${SOURCE_PATIENT_ID}.json`;
+
+async function main() {
+  // TODO: Condense this functionality under a single function and put it on `@metriport/core`, so this can be used both here, and on the Lambda.
```
Comment (Goncharo) [inline packages/utils/src/customer-requests/medical-record-brief-input.ts:25]: tie todo to a specific issue

File: packages/utils/src/customer-requests/medical-record-brief-input.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
+/* eslint-disable */
+// @ts-nocheck
+import * as dotenv from "dotenv";
+dotenv.config();
+// keep that ^ on top
+import { Bundle, Resource } from "@medplum/fhirtypes";
+// import { summarizeFilteredBundleWithAI } from "./medical-record-brief-input";
```
Comment (Goncharo) [inline packages/utils/src/customer-requests/medical-record-brief-input.ts:7]: nit: remove

**Review by Goncharo — APPROVED** [review]
I don't see any show-stoppers here, so approved 🚢 - fine to address everything in a follow up PR

Some things to consider to consider this "production ready" and not alpha:
- make sure to add gating this to a specific cx as part of the release plan
- it would be extremely useful to add metrics around the time it takes to generate a summary, along with the amount of tokens used (cc @Orta21) so we can monitor this feature in prod
- technically, this can take an infinite amount of time, since we're not capping the max number of chunks we feed the model (maybe we should add a timeout per chunk and/or cap the max number of chunks)

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 4 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:74]: Added to #metriport/metriport-internal/issues/2510

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 5 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:371]: Added to #https://github.com/metriport/metriport-internal/issues/2510

File: packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts (hunk 6 above)
Comment (RamilGaripov) [reply packages/core/src/external/aws/lambda-logic/bundle-to-brief.ts:426]: added a comment for now, but will also add TODO to split this up

**Pull request merged**